Doc/c-api/lifecycle.rst

```diff
@@ -0,0 +1,206 @@
+way to break a reference cycle.  The destruction of an object in a cyclic
+isolate would result in a dangling pointer, triggering undefined behavior when
+an object referencing the destroyed object is itself destroyed.
+
+:c:member:`~PyTypeObject.tp_finalize` is not needed to safely destroy a cyclic
+isolate, but its existence makes it easier to design types that behave in a
+sane manner when objects are cleared.  Clearing an object might necessarily
+leave it in a broken state---it might be unsafe to call any of the cleared
+object's methods or access any of its attributes.  With finalization, only
+finalized objects can possibly interact with cleared objects; non-finalized
+objects are guaranteed to interact with only non-cleared (but potentially
+finalized) objects.
+
+To summarize the possible interactions:
+
+* A non-finalized object might have references to or from non-finalized and
+  finalized objects, but not cleared objects.
+* A finalized object might have references to or from non-finalized, finalized,
+  and cleared objects.
+* A cleared object might have references to or from finalized and cleared
+  objects, but not non-finalized objects.
+
+Without any reference cycles, an object can be simply destroyed once its last
+reference is deleted; the finalization and clearing steps are not necessary to
+safely reclaim unused objects.  However, it can be useful to automatically call
+:c:member:`~PyTypeObject.tp_finalize` and :c:member:`~PyTypeObject.tp_clear`
+before destruction anyway because type design is simplified when all objects
+always experience the same series of events regardless of whether they
+participated in a cyclic isolate.  :term:`CPython` currently only calls
+:c:member:`~PyTypeObject.tp_finalize` and :c:member:`~PyTypeObject.tp_clear` as
+needed to destroy a cyclic isolate; this may change in a future version.
+
+
+Functions
+---------
+
+To allocate and free memory, see :ref:`allocating-objects`.
+
+
+.. c:function:: void PyObject_CallFinalizer(PyObject *op)
+
+   Finalizes the object as described in :c:member:`~PyTypeObject.tp_finalize`.
+
+
+.. c:function:: int PyObject_CallFinalizerFromDealloc(PyObject *op)
+
+   Same as :c:func:`PyObject_CallFinalizer` but meant to be called at the
+   beginning of the object's destructor (:c:member:`~PyTypeObject.tp_dealloc`).
+   The object's reference count must already be 0.  If the object's finalizer
+   resurrects the object, this function returns -1; no further destruction
+   should happen.  Otherwise, this function returns 0 and destruction can
+   continue normally.
+
+
+.. rubric:: Footnotes
+
+.. [1] Disclaimer: Some of the details documented in this section may change in
+   a future version of :term:`CPython`.  Care was taken to limit the promises
+   made here without sacrificing information needed to correctly design types
+   and without sacrificing the ability to improve CPython in a
+   backwards-compatible way.  Time will tell whether we were successful.
```

Comment:
If this PR is accepted, we won't need this disclaimer, as breaking documented behaviour would be a backwards-incompatible change.